mev-boost-rs/src/service.rs

```diff
@@ -52,23 +52,44 @@ impl Service {
         Self { host: config.host, port: config.port, relays, network }
     }
 
-    pub async fn run(&self) {
+    /// Spawns a new [`RelayMux`] and [`BlindedBlockProviderServer`] task
+    pub fn spawn(self) -> ServiceHandle {
+        let Self { host, port, relays, network } = self;
         let context: Context = self.network.into();
-        let relays = self.relays.iter().cloned().map(|endpoint| Relay::new(Client::new(endpoint)));
-        let relay_mux = RelayMux::new(relays, Arc::new(context), self.network);
-
-        let mut tasks = vec![];
+        let relays = relays.into_iter().map(|endpoint| Relay::new(Client::new(endpoint)));
+        let relay_mux = RelayMux::new(relays, Arc::new(context), network);
 
         let relay_mux_clone = relay_mux.clone();
-        tasks.push(tokio::spawn(async move {
+        let relayer = tokio::spawn(async move {
             relay_mux_clone.run().await;
-        }));
+        });
+
+        let server = BlindedBlockProviderServer::new(host, port, relay_mux).spawn();
+
+        ServiceHandle { relayer, server }
+    }
+}
+
+/// Contains the handles to spawned [`RelayMux`] and [`BlindedBlockProviderServer`] tasks
+///
+/// This struct is created by the [`Service::spawn`] function
+#[pin_project::pin_project]
+pub struct ServiceHandle {
+    #[pin]
+    relayer: JoinHandle<()>,
+    #[pin]
+    server: JoinHandle<()>,
+}
 
-        let builder_api = BlindedBlockProviderServer::new(self.host, self.port, relay_mux);
-        tasks.push(tokio::spawn(async move {
-            builder_api.run().await;
-        }));
+impl Future for ServiceHandle {
+    type Output = Result<(), JoinError>;
 
-        join_all(tasks).await;
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let relayer = this.relayer.poll(cx);
+        if relayer.is_ready() {
+            return relayer
+        }
+        this.server.poll(cx)
```

Comment:
I guess there is no issue biasing the `relayer` over the `server` in terms of QoS concerns